engine/orchestration/src/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java

```diff
@@ -2868,6 +2849,89 @@ public boolean restartNetwork(final Long networkId, final Account callerAccount,
         }
     }
 
+    @Override
+    public void destroyExpendableRouters(final List<? extends VirtualRouter> routers, final ReservationContext context) throws ResourceUnavailableException {
+        final List<VirtualRouter> remainingRouters = new ArrayList<>();
+        for (final VirtualRouter router : routers) {
+            if (router.getState() == VirtualMachine.State.Stopped ||
+                    router.getState() == VirtualMachine.State.Error ||
+                    router.getState() == VirtualMachine.State.Shutdowned ||
+                    router.getState() == VirtualMachine.State.Unknown) {
+                s_logger.debug("Destroying old router " + router);
+                _routerService.destroyRouter(router.getId(), context.getAccount(), context.getCaller().getId());
+            } else {
+                remainingRouters.add(router);
+            }
+        }
+
+        if (remainingRouters.size() < 2) {
+            return;
+        }
+
+        VirtualRouter backupRouter = null;
+        for (final VirtualRouter router : remainingRouters) {
+            if (router.getRedundantState() == VirtualRouter.RedundantState.BACKUP) {
+                backupRouter = router;
+            }
+        }
+        if (backupRouter == null) {
+            backupRouter = routers.get(routers.size() - 1);
+        }
+        if (backupRouter != null) {
+            _routerService.destroyRouter(backupRouter.getId(), context.getAccount(), context.getCaller().getId());
```

Comment:
it seems this will be called even when a redundant backup router is required in this network. Can you explain how this works?

Reply:
The purpose of removal of expendable router in `destroyExpendableRouters` is to find and destroy VRs of network where destroying has no side effect on the network. We first try to find and destroy bad routers (in non-running states), then we see how many routers are left if they are less than 2 (i.e. not redundant routers) we break/exit the method. Otherwise, we see if any of those two routers are backup router which can be destroyed. This method is shared by both the vpc and non-vpc routers.

Reply:
Is there a reason for this method to be public? it can be called at anytime on a redundant network at which time it would destroy the backup.

Reply:
Yes, the reason is that this common logic is used at two places one where restart of isolated networks are handled and another vpc mgr class that handles the restart of VPCs. @rafaelweingartner advised to refactor the common logic, I did as much I could.

Reply:
got it @rhtyd. I still dread this public method being called from places that do not intent what it was programmed to do. As in this method a backup router will be destroyed no matter what. How about adding some javadoc:
/**
  * This method will first destroy and router in state Stopped, Error, Shutdowned or Unknown
  * Afterthat it will destroy any backup router or, if it can find non the last router in the list
  */

Reply:
Javadocs are useful to keep the history behind these things. Keep in mind that the method is only public (java delimiter access); this means accessible to programmers to re-use its code. It is not something that end-users will have access to.